File: fsharp/sum-of-multiples/SumOfMultiples.fs
module SumOfMultiples

let sum (numbers: int list) (upperBound: int): int =
    let generateMultiples (n: int): int list =
        let inclusiveNumber: int = upperBound / n
        let boundary: int =
            match upperBound % n = 0 with
            | true -> 1
            | false -> 0
        let exclusiveNumber: int = inclusiveNumber - boundary
        List.init exclusiveNumber (fun i -> (i + 1) * n)
    numbers
    |> List.collect generateMultiples
    |> List.distinct
    |> List.sum

File: solutions/scala/sum-of-multiples/1/src/main/scala/SumOfMultiples.scala
object SumOfMultiples
{
  def sum(factors: Set[Int], limit: Int): Int =
  {
    
    def generateMultiples(n: Int): List[Int] =
    {
      val inclusiveNumber: Int = limit / n
      val boundary: Int =
        limit % n == 0 match
        {
          case true => 1
          case false => 0
        }
      val exclusiveNumber: Int = inclusiveNumber - boundary
      List.tabulate(exclusiveNumber)(i => (i + 1) * n)
    }
    factors.flatMap(generateMultiples).sum()
  }
}

